### ISSUE TITLE
Debugging segfaults on workers

### ISSUE THREAD

**Issue by ptooley** [dask/distributed#2243]
I am having awful problems with segfaulting workers and don't really know how to go about debugging.  Typical tracebacks are below, it shows up in the client output as connection errors and workers killed by unknown signals.

As far as I can tell there is no particular part of my code that causes the segfaults as they show up almost randomly in different runs with the same parameters.

This is currently proving showstopping for my use of dask for large array based jobs as I can't do performance or memory use optimisation while my workers are constantly dying.

```
distributed.nanny - WARNING - Worker process 28865 was killed by unknown signal
distributed.nanny - WARNING - Restarting worker
distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:44041
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/comm/tcp.py", line 177, in read
    n_frames = yield stream.read_bytes(8)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
tornado.iostream.StreamClosedError: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 1808, in gather_dep
    who=self.address)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 2808, in get_data_from_worker
    max_connections=max_connections)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/core.py", line 431, in send_recv
    response = yield comm.read(deserializers=deserializers)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/tcp.py", line 198, in read
    convert_stream_closed_error(self, e)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/tcp.py", line 124, in convert_stream_closed_error
    raise CommClosedError("in %s: %s: %s" % (obj, exc.__class__.__name__, exc))
distributed.comm.core.CommClosedError: in <closed TCP>: ConnectionResetError: [Errno 104] Connection reset by peer
distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:44041
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/comm/tcp.py", line 177, in read
    n_frames = yield stream.read_bytes(8)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
tornado.iostream.StreamClosedError: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 1808, in gather_dep
    who=self.address)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 2808, in get_data_from_worker
    max_connections=max_connections)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/core.py", line 431, in send_recv
    response = yield comm.read(deserializers=deserializers)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/tcp.py", line 198, in read
    convert_stream_closed_error(self, e)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/tcp.py", line 124, in convert_stream_closed_error
    raise CommClosedError("in %s: %s: %s" % (obj, exc.__class__.__name__, exc))
distributed.comm.core.CommClosedError: in <closed TCP>: ConnectionResetError: [Errno 104] Connection reset by peer
distributed.nanny - WARNING - Worker process 29375 was killed by unknown signal
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker process 28867 was killed by unknown signal
distributed.nanny - WARNING - Restarting worker
distributed.nanny - WARNING - Worker process 30141 was killed by unknown signal
distributed.nanny - WARNING - Restarting worker
distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:44041
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 186, in connect
    quiet_exceptions=EnvironmentError)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
tornado.util.TimeoutError: Timeout

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 1808, in gather_dep
    who=self.address)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 2802, in get_data_from_worker
    comm = yield rpc.connect(worker)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/core.py", line 725, in connect
    connection_args=self.connection_args)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 195, in connect
    _raise(error)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 178, in _raise
    raise IOError(msg)
OSError: Timed out trying to connect to 'tcp://127.0.0.1:44041' after 3 s: in <distributed.comm.tcp.TCPConnector object at 0x7f9692278ac8>: ConnectionRefusedError: [Errno 111] Connection refused
distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:44041
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 186, in connect
    quiet_exceptions=EnvironmentError)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
tornado.util.TimeoutError: Timeout

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 1808, in gather_dep
    who=self.address)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 2802, in get_data_from_worker
    comm = yield rpc.connect(worker)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/core.py", line 725, in connect
    connection_args=self.connection_args)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 195, in connect
    _raise(error)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 178, in _raise
    raise IOError(msg)
OSError: Timed out trying to connect to 'tcp://127.0.0.1:44041' after 3 s: in <distributed.comm.tcp.TCPConnector object at 0x7f969693c860>: ConnectionRefusedError: [Errno 111] Connection refused
Normalising by 0.982326090335846
distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:44041
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 186, in connect
    quiet_exceptions=EnvironmentError)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
tornado.util.TimeoutError: Timeout

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 1808, in gather_dep
    who=self.address)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/worker.py", line 2802, in get_data_from_worker
    comm = yield rpc.connect(worker)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/core.py", line 725, in connect
    connection_args=self.connection_args)
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
  File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 195, in connect
    _raise(error)
  File "/usr/lib64/python3.6/site-packages/distributed/comm/core.py", line 178, in _raise
    raise IOError(msg)
OSError: Timed out trying to connect to 'tcp://127.0.0.1:44041' after 3 s: in <distributed.comm.tcp.TCPConnector object at 0x7f96c04ce748>: ConnectionRefusedError: [Errno 111] Connection refused
distributed.nanny - WARNING - Worker process 30462 was killed by unknown signal
distributed.nanny - WARNING - Restarting worker
```

```
[10091.246585] python3.6[24625]: segfault at 11 ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10091.336375] traps: python3.6[24623] general protection ip:7efe055d3c1a sp:7ffc8450a070 error:0 in libffi.so.6.0.4[7efe055ce000+7000]
[10099.052996] python3.6[24634]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10100.278125] traps: python3.6[26421] general protection ip:7efe055d3c1a sp:7ffc8450a070 error:0 in libffi.so.6.0.4[7efe055ce000+7000]
[10117.851446] CPU6: Core temperature above threshold, cpu clock throttled (total events = 8009)
[10117.851447] CPU2: Core temperature above threshold, cpu clock throttled (total events = 8009)
[10117.851448] CPU0: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851449] CPU3: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851450] CPU4: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851451] CPU7: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851452] CPU1: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851453] CPU5: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851454] CPU2: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.851455] CPU6: Package temperature above threshold, cpu clock throttled (total events = 8205)
[10117.852348] CPU2: Core temperature/speed normal
[10117.852348] CPU6: Core temperature/speed normal
[10117.852349] CPU5: Package temperature/speed normal
[10117.852350] CPU0: Package temperature/speed normal
[10117.852350] CPU1: Package temperature/speed normal
[10117.852351] CPU4: Package temperature/speed normal
[10117.852351] CPU3: Package temperature/speed normal
[10117.852352] CPU7: Package temperature/speed normal
[10117.852353] CPU6: Package temperature/speed normal
[10117.852354] CPU2: Package temperature/speed normal
[10126.771961] traps: python3.6[27536] general protection ip:7efe055d3c1a sp:7ffc8450a070 error:0 in libffi.so.6.0.4[7efe055ce000+7000]
[10137.567077] python3.6[28620]: segfault at 7efddc13cf4c ip 00007efe055d3c02 sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10137.575706] python3.6[28906]: segfault at a2e737f ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10137.577545] traps: python3.6[28646] general protection ip:7efe055d3c1a sp:7ffc8450a070 error:0 in libffi.so.6.0.4[7efe055ce000+7000]
[10137.780183] traps: python3.6[29251] general protection ip:7efe0a9ae2f5 sp:7ffc84509fb8 error:0 in libpython3.6m.so.1.0[7efe0a942000+242000]
[10137.839648] python3.6[28623]: segfault at 5f5f7c ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10139.729006] python3.6[29862]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10139.731557] python3.6[29828]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10139.737007] python3.6[29880]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10141.038372] python3.6[29982]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10153.693700] python3.6[31084]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10153.768418] python3.6[31136]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10154.875317] python3.6[31169]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10165.830557] python3.6[31762]: segfault at 7efddc13bf4c ip 00007efe055d3c02 sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
[10166.970633] python3.6[32304]: segfault at a ip 00007efe055d3c1a sp 00007ffc8450a070 error 4 in libffi.so.6.0.4[7efe055ce000+7000]
```

**Comment by mrocklin** [comment]
It looks like something is killing your workers:

distributed.nanny - WARNING - Worker process 28865 was killed by unknown signal


Commonly this is because the system you are using to launch Dask (you
mention array jobs, so I suspect that you're using a cluster manager like
SGE, SLURM, Torque, etc..).  Usually this is because they take up more
memory than your cluster manager likes, but there are many possible
reasons.  Unfortuantely it's hard to help without more information.

**Comment by ptooley** [comment]
@mrocklin thanks for the reply.  To be clear this is a Localcluster on my machine, not a managedcluster.  They are definitely segfaults because I am getting the dmesg output telling me about it, but I have no idea what routine is segfaulting and how to find out.

**Comment by mrocklin** [comment]
In that case my best guess is that you're running some code that segfaults
when run in multiple threads or something similar.  You might try running
it in mulitple threads using non-dask tooling (like concurrent.futures) or
you might try running Python within gdb to figure out where the segfault
is.

Generally speaking Dask itself is just a Python application, which unless
you're trying very hard doesn't segfault :)

**Comment by ptooley** [comment]
> Generally speaking Dask itself is just a Python application, which unless you're trying very hard doesn't segfault :)

Well exactly. And my code is pure python as well, making the whole issue very strange indeed.  I shall try attaching gdb to my workers and see what I can find.

**Comment by mrocklin** [comment]
Any update @ptooley ?

**Comment by goraj** [comment]
I have a similar problem as @ptooley, using Client(processes=True) locally my workers segfault/trap in ffi.so as well. Have you found the issue Phil?

**Comment by giantamoeba** [comment]
I had a similar problem. In my case was cloudpickle=0.6.1 to blame. Solved it by downgrading to 0.6.0
